text/0000-chmod.md

```diff
@@ -0,0 +1,62 @@
+# Chmod Buildpack
+
+## Summary
+
+A chmod buildpack that allows users to change directories and files (produced at buildtime) UNIX permissions.
+
+This buildpack would be configured to run last, so that it would be able to change any directories and files produced by previous buildpacks.
+
+## Motivation
+
+Buildstacks and Builders rely on Stacks.
+
+Stacks define build time and runtime base images.
+
+Some stacks do not use the same user / UID between build and run images, [for security reasons](https://github.com/paketo-buildpacks/rfcs/blob/main/text/0045-user-ids.md), since a runtime user should just be able to run a process; not create directories nor install packages.
+
+Unfortunately, some buildpacks do need to create directories and files: they run setup scripts during runtime that fail because they can't write folders nor files, since they were run with a runtime user, who can be restricted with some Stacks. 
+
+Similarly, some scripts that need to be run at runtime are only executable by the build time user, and would need to be executable by the runtime user.
+
+## Detailed Explanation
+
+N/A
+
+## Rationale and Alternatives
+
+We have considered solving the issue at different levels:
+
+* At the stack level: reverting to former stacks that define the same user at build time and runtime.
+
+After all, previous stacks like Ubuntu Bionic did not differentiate the runtime user from the build time user, both being named `cnb` with the UID 1000
+
+We began noticing this issue with the latter stacks (some based on Ubuntu Jammy) that guarantee [a more secure environment](https://github.com/buildpacks/rfcs/blob/main/text/0085-run-uid.md) by having the build time user set to UID 1002, and runtime user set to 1000 at runtime.
+
+* At the buildpack level: [implementing special flags in some buildpacks](https://github.com/paketo-buildpacks/dist-zip/pull/174) that would allow users of those buildpacks to make some folders group writable
+
+But then, we discovered that it was not enough, some other flag would need to be implemented to let the build time user run a different set of scripts; and of course, we would need to replicate this behavior in several other buildpacks packaging applications that set themselves up at runtime.
+
+Allowing the user to keep on using the same secure Stacks, and unmodified buildpacks, using a chmod buildpack at the end of their build seems like the best solution
+
+## Implementation
+
+The chmod buildpack would be added last to builders and would be activated if a specific flag is set.
+
+This flag, an environment variable named `BP_CHMOD_MAPPING` would be configured with a map of folders and their desired owner and permissions.
+
+For example:
+
+`pack build -e BP_CHMOD_MAPPING='{"/workspace":"0750", "/workspace/bin":"0755"}' myimage`
+
```

Comment:
This is the part I am most unsure about & I'll say upfront, my comments here could probably also go under bikeshedding :) 

I agree on using an env variable for input since that is the chosen path for Paketo buildpacks, but I get stuck on the format. Shoving JSON or other structured text into an env variable is a real pain. I've had to do this with other CLI tools, and it is not a great user experience. Quoting get's messed up, and is tricky with different shells (and Windows).

I'm trying to think through other examples of data like this in our buildpacks, but nothing is coming to mind. I have a few ideas though.

1. `BP_CHMOD_MAPPING`, but in a more compact format, like `/workspace:0750:r /workspace/bin:0755` (or maybe `|` as the delimiter). I think JSON is OK in the example, but recursive should be a configurable option so that would make the JSON more complicated/verbose. And quoting is painful.

2. `BP_CHMOD_MAPPING_FILE` points to a file with the mappings in it (maybe one per line, same compact format or JSON, it is more manageable in a file). This could be packaged with the app or it could be mounted into the container as a volume mount. Possibly with a default file like `Permissionsfile` at the root.

3. A binding with type `chmod`. Mentioning to be complete, but I don't think this is the right option, cause it doesn't contain secret data and bindings really should be for secret data.

Perhaps we go with a combination of the 1.) and 2.)? If you have only one or two mappings, an env variable is nice. If you have 20, probably not so much.

I'm also wondering about the MVP here, which is another point to consider. 

Anyway, I'm not specifically suggesting anything here, but hoping to spark a conversation and see what folks think about the format.

Reply:
It's interesting you mention option 2. since it's very similar to what the `native` buildpack [provides](https://github.com/paketo-buildpacks/native-image/blob/main/README.md?plain=1#L27). 

At least, that'd make the UX with paketo similar across buildpacks, so yeah makes sense, 1+2, I'll update the RFC

Reply:
ok it's pushed!